KEP-3698: Multi-Network (#3700)
kubernetes/enhancements: keps/sig-network/3698-multi-network-requirements/README.md

  - [Dependencies](#dependencies)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

One of the main requirements for Kubernetes networking today is to enable
connectivity between `Pods` in a cluster. This facility satisfies a large number of
Kubernetes users, but it isn't sufficient for some important cases:

* applications leveraging different isolated networks exposed through different
  interfaces
* applications leveraging performance-oriented interfaces (e.g. `AF_XDP`,
  `memif`, `SR-IOV`)
* applications requiring support for protocols not yet supported by Kubernetes

Here we propose a solution that enables multiple network interfaces to be
defined in the `Pod` specification so as to provide a standard and clear 
mechanism to implement these more complex configurations.

## Motivation

We want to have a common API allowing us to define a catalog of different
networks in the Kubernetes cluster. It would allow attaching a pod to one or
several networks via a given type of interface depending on its connectivity
or performance needs.

### Goals

> Define user stories and requirements for the Multi-Network effort in Kubernetes.

Reviewer: At today's multi-network sync (May 10th, 2023 - notes available [here](https://docs.google.com/document/d/1pe_0aOsI35BEsQJ-FhFH9Z_pWQcU2uqwAnOx2NIx6OY)) there was a conversation involving several members of the community regarding the desire for this KEP to better define standards for defining additional parameters of a network interface for a `Pod` (perhaps particularly to help support DRA and QoS configurations) which may be implementation-specific, or perhaps generic enough for all implementations.

Personally I agree with the high level goal, but I think we should avoid trying to put new things in core if possible for a variety of long term maintenance reasons. There are other takes on this which I invite conversation to :thread: here about.

Ultimately, I would like to see that conversation continued and resolved before we merge this KEP and the goals here adjusted accordingly.

cc @marquiz (re https://github.com/kubernetes/enhancements/pull/3004)

Reviewer 2: Thanks @shaneutt. I need to read this KEP with thought, but ulitmately I think that DRA (if any) would be in line with requirements. The QoS resources KEP (#3004) is something that could possibly be used to cover some current usage scenarios involving e.g. annotations. But it's good to understand what it is and how it's different.

cc @pohly

Reviewer 3: My thoughts are that there are two places where we might have arbitrary config.
- At the network level. For example, parameters specifying what VLAN ID this network has, or what IP address ranges it is allowed to use.
- At the network interface level. For example, parameters specifying the QoS for a particular interface, or the allowed bandwidth, or the required failure domain to allow redundancy.

These are all going to be implementation dependent, and we shouldn't put them into core Kubernetes. The easiest way to do this is to have a CR with whatever specific implementation specific horrors are required in it, and allow the standard K8s objects to have a reference to that CR (with it being entirely up to the implementation what to expect in the CR and what to do with it).

There is already a reference to a CR in the PodNetwork resource. My takeaway from the discussion was that the PodNetworkAttachment resource probably needs exactly the same capability.

I'm open to some common fields making their way into the PodNetworkAttachment (say MAC address or even IP as well as interface name or similar), but I can see that turning into something pretty ugly pretty fast if we aren't careful, so shoving as much as possible into a CR seems safer. I'm also open to this being "something we add in a later phase, but agree we expect to do later". Finally, I think the alternative is some kind of annotation model, where we optionally list interfaces and references to their config in annotations in the pod, which is plausible; a bit ugly but also less intrusive to core K8s.

Reviewer 4: As this is a requirements KEP, we should try to avoid blocking it on other KEPs unless we think it is fundamentally a non-starter for the community. The cost paid for work on refining requirements is community time rather than long-term tech debt.

### Non-Goals

Define API and implementation.

## Proposal

### Personas

#### Cluster operator

This role will be responsible for defining and managing PodNetworks (see below)
that properly describe the infrastructure available for the cluster, namespace
and workloads. This persona can define which users can “attach” to a specific
PodNetwork.

#### Application developer

**Application developer** is the consumer of PodNetwork via referencing them in
their workloads. Application developers usually will not create or remove the
PodNetwork on their own.

### Terminology

* **PodNetwork** - name of the API object that will represent a network in
  Kubernetes cluster

> * **Cluster Default PodNetwork** - This is the initial cluster-wide PodNetwork

Reviewer 4: I think we should remove "cluster creation" and life cycle as this could change after create?

Suggest shorten to:

This is the PodNetwork attached to a Pod when no additional...

Author: done

  provided during cluster creation that is available to the Pod when no
  additional networking configuration is provided in Pod spec.

> * **Primary PodNetwork** - This is the PodNetwork inside the Pod which interface

Reviewer 4: We should be specific about the cardinality relationship between PodNetwork and interfaces. Is it always 1-1? If not, we will need to be very careful whenever one is mentioned vs another.

Reviewer 5: I assume we will probably land up with either a scope field in the API or different APIs (Kinds) for each of these use-cases ? Wondering if the following nomenclature pattern is more intuitive.

1) Cluster Default PodNetwork
2) Namespace Default PodNetwork (For future expansion, note Default is not the name of the namespace, this is the default network associated with a namespace)
3) Default PodNetwork (instead of primary network, this is the one assigned to the pod, either by policy or by the user picking one).

Author: @bowei The cardinality relationship is going to be defined in the Phase I. I dont want to mention that in this phase.

@pvenu-Google using "Default PodNetwork" when we already have "Cluster Default PodNetwork" will not be distinguishable, most folks will keep skipping the "Cluster" word which will lead to confusion. And we want to keep the "Default" PodNetwork for what it is now to align with the "default" namespace.

  is used for the default gateway.
 

### User Stories

All user stories represent the type of use cases the multi-networking API should
be able to support. References to technologies or exact products does not
indicate that this API will directly support them. The set of requirements,
defined out of these use cases, are the final indicator of what will be covered
by this API and effort.

> 

Reviewer 4: Suggest:

User stories can be overlapping in some areas. We want to document the concrete use cases from a end-user standpoint -- they may (and should!) boil down to a few common primitives.

Author: updated

availability (e.g. devicePlugin or some other way).

<p align="center">
  <img src="mn-story-2.png?raw=true" alt="multi-network story 2 HW interface"/>
</p>

#### Story #3
I have implemented my Kubernetes cluster networking using a virtual switch. In
this implementation I am capable of creating isolated Networks. I need a means
to express to which Network my workloads connect to.

<p align="center">
  <img src="mn-story-3.png?raw=true" alt="multi-network story 3 virtual switch isolation"/>
</p>

#### Story #4
As a Virtual Machine -based compute platform provider that I run on top of
Kubernetes and Kubevirt I require multi-tenancy. The isolation has to be
achieved on Layer-2 for security reasons.

<p align="center">
  <img src="mn-story-4.png?raw=true" alt="multi-network story 4 Kubevirt VMs"/>
</p>

#### Story #5
As a platform operator I need to connect my on-premise networks to my workload
Pods. I need to have the ability to represent these networks in my Kubernetes
cluster in such a way that I can easily use them in my workloads.

<p align="center">
  <img src="mn-story-5.png?raw=true" alt="multi-network story 5 On-Premise Network representation"/>
</p>

#### Story #6

> As a Kubernetes cluster operator I wish to isolate workloads based on

Reviewer 4: I think this is an important point which needs to be called out more -- which is the role-based ACLs for network configuration that /may/ be independent of people who run the workloads.

Author: Not sure I fully understand, but we do call out "Personas" above and how those relate to PodNetwork handling.

Pod network namespace without any restrictions. I am aware that by doing this I
might break established contracts for the kubernetes features.

#### Story #8
As a Virtual Machine -based compute platform provider that I run on top of
Kubernetes and Kubevirt I need the ability to add/remove Network Interfaces to
existing VMs without re-creating the VM. This would not be applicable for the
Primary PodNetwork.

<p align="center">
  <img src="mn-story-8.png?raw=true" alt="multi-network story 8 network interface hot-plug"/>
</p>

#### Story #9
As an infrastructure administrator I wish to support IPv6 SLAAC in the
Kubernetes Pods. SLAAC's nature is to dynamically assign IP addresses and change
them over time. Those should be reflected in Pod status. This would not be
applicable for the Primary PodNetwork.

### Requirements

Below requirements are divided into phases. Each phase will produce a separate
KEP with detailed design for specified set of requirements. Each of this KEP
will have to take all of the requirements, from each phase, into consideration
when created.

#### Phase I (base API and reference in Pod)

> 1. This effort shall not change the behavior of Today existing clusters

Reviewer 4: Suggest:

Multi-network should not change the behavior of existing clusters where no multi-network constructs are configured/used.

Author: updated

> 2. We need an API object representing network, which a Pod, a Service or any

Reviewer 4: Suggest:

Multi-network needs to be represented explicitly as an API construct in Kubernetes as it fundamentally ties in with other APIs in the Kubernetes networking model.

Author: updated

> other feature that wishes to support Multi-Network capability could reference

Author: Today's implementation are creating CRDs, that have to be referenced via annotations or other in-direct means. Annotation is not meant to define configuration of an object, additionally there is no validation of the user input.
On top of that every implementation has its own CRDs and way of configuration, so there is no one standard.

> other feature that wishes to support Multi-Network capability could reference

Reviewer 6: > Today's implementation are creating CRDs, that have to be referenced via annotations or other in-direct means.

Some network implementation uses CRDs, but not all. For example, flannel does not use CRDs. In addition, each implementation has several CRDs to specify network but it is depends on the implementation. Some implementation use a CRD for 'network', but some implementation use CRDs for 'network IPAM' and 'network transport (i.e. SDN configuration)'.

 Your design assumes that one network object is required for everyone, for each implementation, but actually it is not, I guess. So please clarify and give a reason why we need the object. Currently the reason seems to be ambiguous.

> other feature that wishes to support Multi-Network capability could reference

Author: The main reason is to have a handle/representation of "network" Pod connects/attaches to. This does NOT exists Today. This is mentioned in the Summary and Motivation, maybe it is not sufficient. I will add more there.

   * We will introduce an PodNetwork API object that represents the existing 
     infrastructure’s networking
3. PodNetwork is decoupled from the infrastructure’s network

> definition/description created by the cluster operator. It functions as a handle

Reviewer 4: incomplete sentence?

suggest:

This functions as a handle that decouples the implementation (i.e. infrastructure/network admin's) from the consumer (cluster user). This abstraction follows the Kubernetes model of decoupling intent from implementation.

Author: updated.

4. PodNetwork shall not define any implementation specific parameters in its
specification

> 5. PodNetwork shall provide cluster generic options only

Reviewer 4: I think there needs to be more specificity on what "cluster generic options only" means. This is in contrast to?

Author: This is in contrast to implementation/technology -specific parameters. Maybe "core kubernetes options only" ?

6. PodNetwork can reference to implementation-specific parameters

> 7. PodNetwork is the object Application developer will reference in their workloads

Reviewer 4: in their workflows (e.g. Pod, Service)

> 8. Cluster Default PodNetwork is the PodNetwork the cluster has been created with

Reviewer 4: Is this strictly a requirement? This implies some amount of immutability.
It would be good to understand if there is some lifecycle associated with this that is implied.

Author: We had a discussion with the community around this, and yes we have a path for mutation of the "Cluster Default PodNetwork", updated 9, to mention it HAS to be replaceable.
As to the creation part in this point, I want to still keep the  requirement that there has to be a default PodNetwork defined before the cluster is Ready when created.

9. Cluster Default PodNetwork cannot be changed or removed 
10. Pods can reference one or more PodNetworks when trying to attach to them
    * This list is explicit and all the PodNetworks that has to be available in
      the Pod has to be listed
11. Workloads access to PodNetwork is controlled via RBAC configuration
12. When Pods does not specify any reference to PodNetwork it connects to Cluster
Default PodNetwork (network the cluster has been created with)
13. Pod shall be able to provide additional configuration on how it attaches to
a PodNetwork

> 14. Every Pod connected to specific PodNetwork must have connectivity within

Reviewer 4: Do we want to define what "connectivity" means specifically.

We should state explicitly that we are probably never going to mandate a requirement (think K8s conformance) that two networks have or don't have connectivity to each other because it is out of scope of a given K8s cluster -- e.g. someone can always insert a connection there outside of the cluster infrastructure and we can't rely or enforce this easily.

Author: I dont think we specify what "connectivity" means in Today's Kubernetes network requirement neither. Additionally, PodNetwork's implementations might be done with different technologies which will make it hard to have one definition for it.

I have created separate requirement for the X-connectivity. keeping it formulated as a requirement.

that network across the Cluster
    * A Pod connected to specific PodNetwork may or may not have cross
      connectivity between different PodNetworks
15. Pods attached to PodNetworks are connected to each other in a manner 

> defined by the PodNetwork implementer

Reviewer 4: implementer => implementation

Author: updated

> 16. Basic PodNetwork Interface information for each attachment will be exposed

Reviewer 4: Basic -> implementation-agnostic

Author: updated

to runtime Pod (via e.g. environment variables, downward API etc.)

#### Phase II (scheduler, kubelet and API probing for PodNetwork)
17. PodNetworks can be selectively (per Node) available in the Cluster. This
does NOT apply to the Cluster Default PodNetwork

> 18. Kubelet network-based probing can be optionally enabled for the non- Cluster

Reviewer 4: Does this refer to Kubelet-based healthchecks? Would help to link to the feature more explicitly.

Author: This refers to all the possible network-based communication kubelet has with the Pod via Today's eth0. Are healthchecks the only thing?

Default PodNetwork attachments to Pod

> 19. Kubernetes API can be optionally accessible via the non- Cluster Default

Author: updated.

> 19. Kubernetes API can be optionally accessible via the non- Cluster Default

Author: On second thought, this one is about e.g. webhook connectivity on non- Cluster Default PodNetwork. The general Service support is in next requirement.

    - [Story #6](#story-6)
    - [Story #7](#story-7)
    - [Story #8](#story-8)
    - [Story #9](#story-9)
  - [Requirements](#requirements)
    - [Phase I (base API and reference in Pod)](#phase-i-base-api-and-reference-in-pod)
    - [Phase II (scheduler, kubelet and API probing for PodNetwork)](#phase-ii-scheduler-kubelet-and-api-probing-for-podnetwork)
    - [Phase III (basic Kubernetes features integration)](#phase-iii-basic-kubernetes-features-integration)
    - [Phase IV (extended functionality)](#phase-iv-extended-functionality)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
    - [Beta](#beta)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

> Today Kubernetes Networking is very straightforward and easy to achieve. Main
> requirement is to enable connectivity between the Pods in the cluster. That
> simple approach satisfies most of Kubernetes customers, but it is not sufficient
> for cases where more complex networking is required:
> * some applications leverage different isolated networks, exposed through
>   different interfaces
> * some leverage performance oriented interfaces (e.g. AF_XDP, memif, SR-IOV),
>   besides the regular management interface
> * others require support for specific protocols not yet supported in Kubernetes
> 
> For such requirements we need a solution that allows a user to express this in
> the pod specification. This can be done by an external solution leveraging
> annotations, but having it incorporated in Kubernetes would be a much cleaner
> and safer approach that would allow better compatibility and consistency for
> workloads with these needs.
> 
> This KEP is an entry level document for the whole Multi-Network endeavor. Here
> we will define a set of requirements to follow. Additionally, we will introduce
> a phased approach where each phase will have its own KEP with detailed design.

Reviewer: I had some thoughts about shortening the summary up a bit:

```suggestion
One of the main requirements for Kubernetes networking today is to enable
connectivity between `Pods` in a cluster. This facility satisfies a large number of
Kubernetes users, but it isn't sufficient for some important cases:

* applications leveraging different isolated networks exposed through different
  interfaces
* applications leveraging performance-oriented interfaces (e.g. `AF_XDP`,
  `memif`, `SR-IOV`)
* applications requiring support for protocols not yet supported by Kubernetes

Here we propose a solution that enables multiple network interfaces to be
defined in the `Pod` specification so as to provide a standard and clear 
mechanism to implement these more complex configurations.
```

Only if you like them, if you don't no worries please feel free to resolve this comment at your discretion.

## Motivation

We want to have a common API allowing us to define a catalog of different
networks in the Kubernetes cluster. It would allow attaching a pod to one or
several networks via a given type of interface depending on its connectivity
or performance needs.

### Goals

Define user stories and requirements for the Multi-Network effort in Kubernetes.

### Non-Goals

Define API and implementation.

## Proposal

### Personas

#### Cluster operator

This role will be responsible for defining and managing “Networks” (name TBD,

> referred as “Object” in the Requirement section) that properly describe the

Reviewer 7: Is it not referred to as "PodNetwork"?

```suggestion
referred as “PodNetwork” in the Requirement section) that properly describe the
```

infrastructure available for the cluster, namespace and workloads. This persona
can define which users can “attach” to a specific “Network”.

#### Application developer

**Application developer** is the consumer of Networks (name TBD, referred as

> “Object” in the Requirement section) via referencing them in their workloads.

Reviewer 7: Likewise, I think it is referred to as "PodNetwork".

```suggestion
“PodNetwork” in the Requirement section) via referencing them in their workloads.
```

> # KEP-2829: Migrate Gateway API to k8s.io Group

Author: ha, missed a spot :)

and safer approach that would allow better compatibility and consistency for
workloads with these needs.

This KEP is an entry level document for the whole Multi-Network endeavor. Here
we will define a set of requirements to follow. Additionally, we will introduce
a phased approach where each phase will have its own KEP with detailed design.

## Motivation

We want to have a common API allowing us to define a catalog of different
networks in the Kubernetes cluster. It would allow attaching a pod to one or
several networks via a given type of interface depending on its connectivity
or performance needs.

### Goals

Define user stories and requirements for the Multi-Network effort in Kubernetes.

### Non-Goals

Define API and implementation.

## Proposal

### Personas

#### Network Administrator

With this proposal we will introduce a new “persona”, called **Network
Administrator**. This role will be responsible for defining and managing
“Networks” (name TBD, referred as “Object” in the Requirement section) that
properly describe the infrastructure available for the cluster, namespace and
workloads. This persona can define which users can “attach” to a specific

> “Network”.

Author: Fair, we do have "Cluster operator" in mind here, will use that

#### User

User is the consumer of Networks (name TBD, referred as “Object” in the
Requirement section) via referencing them in their workloads. Users usually will

> not create or remove the Network on their own.

Author: Looked at Gateway API, and will rename this to "Application developer", this is what I have in mind here.

### Terminology

* **Default network** - This is the initial cluster-wide Pod networking provided
  during cluster creation that is available to the Pod when no additional
  networking configuration is provided.
* **Primary network** - This is the network inside the Pod which interface is

>   used for the default gateway.

Author: The additional networks can be "cluster-wide" and/or "Traditional Kubernetes networks". Naming is always tricky, do you have other recommendations?

>   used for the default gateway.

Reviewer 8: I personally think you cannot avoid the "default" term in both scenarios due to historical semantics of networking. The "default network" could be "default instance" or other term based on naming of the object.
The "primary network" could be referred to as "primary attachment" where here the logic is to specify to which object the Pod "primary" interface must be attached

### User Stories

All user stories represent the type of use cases the multi-networking API should
be able to support. References to technologies or exact products does not
indicate that this API will directly support them. The set of requirements,
defined out of these use cases, are the final indicator of what will be covered
by this API and effort.

#### Story #1

> As a Cloud Native Network Function (CNF) vendor I require an additional

Author: I was corrected on this in my initial doc, and will follow https://github.com/cncf/telecom-user-group/blob/master/whitepaper/cloud_native_thinking_for_telecommunications.md#1.4.

interface to be provisioned into the Kubernetes Pod. Each of these interfaces
has to be in an isolated network for regulatory compliance. The isolation has to

> be done on a Layer-2.

Reviewer 9: @danwinship Again: concur.  I can think of several different requirements for which folks will sometimes articulate a solution shaped like this, but they almost never at the end of the day are actually *rooted* in this requirement.  This is not a requirement, its a design solution for a problem... and the problem is unstated.

> be done on a Layer-2.

Reviewer 8: There is 2 angles to this dilemma. 

Granted: There is a need to specify the root requirement that drives the need for a different networking construct without going directly in the solutionning. 

However: There is a need to not go and challenge every requirement but allow a native support model.

<p align="center">
  <img src="mn-story-1.png?raw=true" alt="multi-network story 1 network L2 isolation"/>
</p>

> #### Story #1a

Reviewer 10: @k8s-ci-robot

As a Cloud NativeNetwork Function (CNF) vendor I require an additional interface
to be provisioned into the Kubernetes Pod. Each of these interfaces has to be in
an isolated network for regulatory compliance. The isolation has to be done on a

> Layer-3.

Reviewer 9: Again... this sounds like a solution in search of a problem... but it also sounds like there are real problems underneath... could we get those?

> Layer-3.

Reviewer 8: I think emphasis for all scenarios 1a, 1b, 2, 3, etc. Is not really the details but rather the understanding that the multi-network object cannot be built around a single minded view of the approach (ie there shall be another interface and it shall be layer attached)

<p align="center">
  <img src="mn-story-1a.png?raw=true" alt="multi-network story 1a network L3 isolation"/>
</p>

#### Story #2
As a Cloud Native Network Function (CNF) vendor I require a HW-based interface
(e.g. SRIOV VF) to be provisioned to my workload Pod. I need to leverage that HW
for performance purposes (high bandwidth, low latency), that my user-space
application (e.g. DPDK-based) can use. The VF will not use the standard netdev
kernel module. The Pod’s scheduling to the nodes should be based on hardware

> availability (e.g. devicePlugin or some other way).

Reviewer 9: Usually, when you dig into this, what you discover is that you have a workload that requires some sort of special treatment from the physical network and a fast path to it.

The problem is that this requirement is entirely mis-focused.   Literally nobody wants an SRIOV VF.  What they want is a fast path to the physical network doing a specific set of things for them.  

You can tell this because 
1. If you offer someone a VF leads to the physical network *mishandling* the packets pushed into the VF, they will be very unhappy.
2. If you can't offer them a SRIOV VF because your hardware only supports SIOV (nexts technology coming after SRIOV) they will be very unhappy (even if you could have given them a vfio for the SIOV that is connected to the proper physical network treatment)
3. If you offer them vfio for the VF when they needed a kernel interface, they will be unhappy.

All of this becomes even more crucial when you are scheduling the Pod.  If I schedule the Pod to a Node that just has SRIOV VFs available, but its NICs lack the required capabilities, or cannot provide a fast path to the required sort of treatment, this all breaks down.

There *are* important sets of requirements in here... but I would recommend thinking them through outside of the 'interface' 'network' paradigm being applied here to get to the real root of the need.

I'd suggest rethinking the requirement in terms of:

1.  What Network Service does the user want from the physical network?
2. How do they want to consume it?  Via what mechanism? (kernel interface, rdma, vfio, etc)
3. What capabilities do they need from the fast path connection to that Network Service?

> availability (e.g. devicePlugin or some other way).

Reviewer: I agree with Dan that this is a particularly compelling user story. As for SRIOV VF I get that this was used as an example and so focusing in on it specifically might be distracting, but I also think it might be worth taking a couple seconds to add a couple more examples just to moot the point.

<p align="center">
  <img src="mn-story-2.png?raw=true" alt="multi-network story 2 HW interface"/>
</p>

#### Story #3
I have implemented my Kubernetes cluster networking using a virtual switch. In
this implementation I am capable of creating isolated Networks. I need a means

> to express to which Network my workloads connect to.

Reviewer 9: @danwinship Agreed.  This requirement feels like it would be covered by a properly written 'traffic segregation' requirement.

<p align="center">
  <img src="mn-story-3.png?raw=true" alt="multi-network story 3 virtual switch isolation"/>
</p>

#### Story #4
As a Virtual Machine -based compute platform provider that I run on top of
Kubernetes and Kubevirt I require multi-tenancy. The isolation has to be
achieved on Layer-2 for security reasons.

> 

Reviewer 9: Why does this have to be achieved at L2 for security reasons?  That's an oddly specific requirement.  What security reasons?

This is particularly worrisome as K8s carefully avoids having *any* L2 concepts in it.  It would be unfortunate to re-introduce them for such sparse requirements.

<p align="center">
  <img src="mn-story-4.png?raw=true" alt="multi-network story 4 Kubevirt VMs"/>
</p>

#### Story #5
As a platform operator I need to connect my on-premise networks to my workload
Pods. I need to have the ability to represent these networks in my Kubernetes
cluster in such a way that I can easily use them in my workloads.

> 

Reviewer 9: Could we get more clarity on this case?  There are lots of ways of achieving this already using BGP in many existing CNI plugins.  Is this perhaps another traffic segregation case?

Reviewer 8: Challenge ... most existing CNI plugins do it by specific CRDs which do not necessarily coexist well. Purpose is to offer a solution where K8S could support such implementations without specific CRDs.

<p align="center">
  <img src="mn-story-5.png?raw=true" alt="multi-network story 5 On-Premise Network representation"/>
</p>

#### Story #6
As a Kubernetes cluster administrator I wish to isolate workloads based on

> namespaces and network access via assigning different default Network to a

Reviewer 6: Could you please clarify 'default Network' in the story more because the term 'default network' seems to be a bit different from 'default network' in 'Terminology' section.

In the 'Terminology' section, default network is the initial cluster-wide Pod networking, hence the term, 'different default Network', seems to be weird to me.

Author: Yeah, here it is about the "is available to the Pod when no additional networking configuration is provided" part of that definition so it is partially right, but incorrect in the part you mention. Should we just call this "namespace-default" Network?
"namespace-default" Network - network to which Pod is connected to when no networking configuration is specified in Pod, currently it is always the cluster default Network

Reviewer 6: I agree that we should define new term in 'Terminology'.

> "namespace-default" Network - network to which Pod is connected to when no networking configuration is specified in Pod, currently it is always the cluster default Network

Have another question. Based on your description, 'namespace-default' Network is added if 'no network configuration'. Does this mean this network is only added if no network is configured? (I mean that this network is not added when some network is configured)?

In addition, let me clarify that 'namespace-default' network is unique in each namespace or not (to clarify why you add 'namespace' in the term).

Author: > Does this mean this network is only added if no network is configured? (I mean that this network is not added when some network is configured)?

That is correct, there is no assumption, only explicit list of what to attach to, see req 10 below.

> In addition, let me clarify that 'namespace-default' network is unique in each namespace or not (to clarify why you add 'namespace' in the term).

This story is about adding such capability, having different default per namespace. Initially there will be only 1 cluster-wide default. This story is in phase IV, so probably we would return to this in next year or so.

given Namespace. I do not want the tenants to change their manifests for that purpose.
Those workloads should have the same level of Kubernetes functionality:
Services, NetworkPolicies, access to Kubernetes API. I wish to support
“profiles” for Namespace, where I can not only change default Network,
but define a set of Networks assigned to given Namespace that Pods created in
that NS are automatically attached to.

> 

Reviewer 9: OK... so there's a desire here to isolate workloads (traffic segregation).  This time by namespace.  It sounds like in this case there's a desire to have a segregated primary network by namespace, but its unclear to me from the requirement as written why this would imply a 'set of Networks'

It feels like this may be smashing two requirements together:

1.  A desire to have per-namespace traffic segmentation for primary networks.  If so this opens interesting questions about how that interacts with the rest of K8s networking
2. A desire to add a set of networks by namespace to a Pod... but its unclear *why* in this requirement.

<p align="center">
  <img src="mn-story-6.png?raw=true" alt="multi-network story 6 namespace networks profiles"/>
</p>

#### Story #7
As a “Power User” with admin privileges I wish to have the ability to modify my
Pod network namespace without any restrictions. I am aware that by doing this I
might break established contracts for the kubernetes features.

#### Story #8
As a Virtual Machine -based compute platform provider that I run on top of
Kubernetes and Kubevirt I need the ability to add/remove Network Interfaces to
existing VMs without re-creating the VM. This would not be applicable for the
Primary Interface of the Pod.

<p align="center">
  <img src="mn-story-8.png?raw=true" alt="multi-network story 8 network interface hot-plug"/>
</p>

#### Story #9
As an infrastructure administrator I wish to support IPv6 SLAAC in the
Kubernetes Pods. SLAAC's nature is to dynamically assign IP addresses and change
them over time. Those should be reflected in Pod status. This would not be

> applicable for the Primary Interface of the Pod.

Author: You presented different example, but I think it is the same as the above. If I change from one IP to another, or from empty to a new one, I would consider same thing. Lastly, the below req 24 covers this capability.

### Requirements

Below requirements are divided into phases. Each phase will produce a separate
KEP with detailed design for specified set of requirements. Each of this KEP
will have to take all of the requirements, from each phase, into consideration
when created.

#### Phase I (base API and reference in Pod)
1. This effort shall not change the behavior of Today existing clusters
2. We need to introduce an “Object” that represents the existing 

> infrastructure’s networking

Reviewer 9: Question: Why?  Why do we need to introduce an Object here?  This feels more like the 'solution' side of things than the 'requirement' side of things.  

As engineers, we *all* have a serious tendency to decide on a solution and then decide that solution is a requirement.  We all do it (I more than most).  This *screams* that here.  Lets get the real requirements underpinning this proposed solution.

3. “Object” is decoupled from the network definition/description created by the
network administrator
4. The “Object” shall not define any implementation specific parameters in that

> object

Reviewer 9: I am also confused.

5. “Object” shall provide option to define:
   * IPAM mode: external/internal

>    * List of route prefixes - optional and not forced on the implementations

Reviewer 9: In addition to @danwinship 's points here (which I support)... how are we handling or preventing IPAM that conflicts with the Cluster/Service CIDRs and routes that highjack Cluster/Service CIDRs?

>    * List of route prefixes - optional and not forced on the implementations

Reviewer 9: On other point to make: by putting this 'Object' in the K8s API Server, it intrinsically is moved out of the world of the people actually administering the network and into the world of people administering the cluster.   Generally speaking in my experience this will work some of the time, break horribly some of the time, and work with a lot of pain for all involved some of the time, depending on the interactions between the K8s administrators and the folks administering the physical network involved.

Having the flexibility to move this point of control around is going to be crucial, and that argues against it necessarily being *required* to live in the K8s API server.

>    * List of route prefixes - optional and not forced on the implementations

Author: refactored.

> 6. “Object” can reference to implementation-specific parameters

Reviewer 11: +1

7. “Object” is the consumer-facing declaration for workloads
8. Cluster “Default” “Object” is the network the cluster has been created with
9. Cluster “Default” “Object” cannot be changed or removed
10. Pods shall reference the “object” when trying to attach to the specific
networks
    * Pods has to specify all the “object” that it wishes to attach to,
      including the Cluster “Default” “Object”
11. Workloads can access “Object” when “attach” RBAC exists at the attachment time
    * “attach” RBAC is new verb to allow referencing API objects
12. The Pod reference to a Networks is optional and when NOT specified, Pod

> connects to “Default” “Object” (network the cluster has been created with)

Reviewer 6: Based on the discussion, https://github.com/kubernetes/enhancements/pull/3700#pullrequestreview-1235360826, pod may NOT connect to "Default" network and it connect to "namespace-default" Network, instead of "Default". 

This may changes current behavior and it violates the first requirement, `1. This effort shall not change the behavior of Today existing clusters`, because existing cluster connects all pod in cluster network (I believe that is "Default" network in your KEP).

So could you please refine them to be consistent? (change requirement 12 and "namespace-deafult" network, or just remove requirement 1)

13. Pod shall be able to provide additional configuration on how it attaches to
a network
    * Identify what “Object” is the Primary network
    * Optional parameters: MAC address, IP address, speed, MTU, interface name etc.

> 14. Every Pod connected to specific network (represented by the “Object”) must

Reviewer 12: Is this a necessary limitation for all "Objects"? Perhaps a network is meant only for north-south traffic. I wonder if this could be left to be defined by the implementation, similarly to the next point.

Author: I think that would be too open. Today's K8s network has exactly this requirement, this way we standardize across all the networks.

Do you have a use case for such situation?

have connectivity within that network between each other within the Cluster
    * A Pod connected to specific network (represented by the “Object”) may or
      may not have cross connectivity between different networks (represented by
      the “Object”)
15. Pods attached to a network are connected to each other in a manner defined
by the “Object” implementation

> 16. Basic network Interface information for each attachment will be exposed to

Reviewer 12: Is there a comparable downward API today? If not, I wonder if this could be deferred to a later phase. I'm concerned about implications of hot-plug or dynamic IPAM on this. It would also make this first phase a little leaner.

Author: I dont think there is one. Fair point on maybe moving it to later phase. What is your concern in regards to hot-plug or dynamic IPAM?

runtime Pod (via e.g. environment variables, downward API etc.)

> #### Phase II (scheduler, kubelet and API probing)

Reviewer 12: Would QoS and resource handling fit here? Since this phase is concerned with scheduler, maybe it should cover resource management too - if network would be in the same category as CPU or memory, it would suggest using the same scheduling/pooling mechanism

Author: I think QoS and "network-resourcing" as a whole would be a new requirement. I would consider this as an future extension.

Reviewer 13: Shouldn't this be "implementation specific"?  aka something your CNI plugin(s) deal with?
